cdap-api/src/main/java/io/cdap/cdap/api/app/ApplicationConfigurer.java

```diff
@@ -119,4 +121,25 @@ default void emitMetadata(Metadata metadata, MetadataScope scope) {
    * @return The {@link TriggerFactory} used to get triggers
    */
   TriggerFactory getTriggerFactory();
+
+  /**
+   * Return the runtime configurer that contains the runtime arguments and provides access for other runtime
+   * functionalities. This is used for the app to provide additional information for the newly generated app spec
+   * before each program run. This method will return null when the app initially gets deployed.
+   *
+   * @return the runtime configurer, or null if this is the initial deploy time.
+   */
+  @Nullable
+  default RuntimeConfigurer getRuntimeConfigurer() {
+    return null;
+  }
+
+  /**
+   * Return the namespace the app is deployed.
+   *
+   * @return the namespace the app is deployed
+   */
+  default String getDeployedNamespace() {
+    throw new UnsupportedOperationException("Getting deployed namespace is not supported");
```

Comment:
Because this class is in api module, simply add a method without implementation will cause all class which uses api fail to compile, so add this default method to avoid that, this is our typical way to add new method to api module